### PR TITLE
feat(Progress): added helper text

### LINKED ISSUE #8240: Progress - Add support for helper text
The ability to support helper text on the Progress component is needed to support this feature request here: https://github.com/patternfly/patternfly-react/issues/8238

### PR DESCRIPTION
**Pull request by wise-king-sullyman**
**What**: Closes #8240 

**Additional issues**:

Convenience link to the added example: https://patternfly-react-pr-8307.surge.sh/components/progress/#helper-text

### CODE REVIEW

**Comment by wise-king-sullyman** [comment]
I would appreciate thoughts on (in particular):
1. The example for this new feature. I debated between doing something like what I ended up with or just showing with one variant (probably `danger`) and would be happy with either approach, or of course something different!
2. The API for helper text variant. I could refactor to allow the helper variant to be controlled independently of the progress variant, or only support `error` based helper text since that was all that was originally requested.

**Review by thatblindgeye — COMMENTED** [review]
This looks nice! Had a couple comments below.

Regarding the way the example is setup, I do like the approach you took if we want to allow users to see all of the valid variants of the helper text without having to type it in in the code editor. It can be a nice middle-ground between having a separate example for each variant and only showing one variant.

**Review by mcoker — COMMENTED** [review]
LGTM but I'm curious if the user can use more complex helper text there if they wanted. Like add an icon or have multiple helper text items if the progress does something like runs through multiple jobs and logs status helper text per job (similar helper text output as the [password strength demo](https://www.patternfly.org/v4/demos/password-strength)). I'm not sure if that's a use case, but I would likely expect the user to be able to optionally pass and manage their own `<HelperText>` here if they have a more complex use case.

File: packages/react-core/src/components/Progress/ProgressHelperText.tsx
Comment (wise-king-sullyman) [reply packages/react-core/src/components/Progress/ProgressHelperText.tsx:27]: Yeah I could see situations where it may not be needed so I'm not against adding a prop for that.

Do you have a preference for defaulting it to be applied or not? I feel like more use cases would benefit from it being a live region than not so I _think_ I lean towards having the default behavior be to apply it.

File: packages/react-core/src/components/Progress/ProgressHelperText.tsx
Comment (wise-king-sullyman) [reply packages/react-core/src/components/Progress/ProgressHelperText.tsx:15]: I wouldn't be opposed to doing so. I originally had spread props but I removed them because it didn't feel like this component made sense to expose publicly, and given our current use case they didn't seem to add any value.

File: packages/react-core/src/components/Progress/ProgressHelperText.tsx
Comment (thatblindgeye) [reply packages/react-core/src/components/Progress/ProgressHelperText.tsx:27]: I think the only reason I might suggest it _not_ being applied by default is to follow how AlertGroup and HelperText both default to their live regions not being applied by default (ignoring cases where an `aria-live` is hardcoded rather than controlled via prop, like ClipboardCopy's internal Tooltip).

**Comment by wise-king-sullyman** [comment]
@mcoker we could definitely adjust this component to be more composable like that so that consumers can pass in their own `HelperText`. My only question there is how we would handle consumers passing *other* components in, would there be any concerns about that from the core side?

**Comment by mcoker** [comment]
@wise-king-sullyman it shouldn't, it's just a container for the help text - what goes in it is up to the user. They could place their own content there if they want, or some other component that serves a similar purpose if we introduce one. That said, I think it would be good to get other opinions on how to best handle this, since we'll likely do the same thing in other components that we're doing here.

File: packages/react-core/src/components/Progress/ProgressHelperText.tsx
Comment (wise-king-sullyman) [reply packages/react-core/src/components/Progress/ProgressHelperText.tsx]: In the most recent changes I removed the default `isLiveRegion`. If a consumer wishes to enable that behavior they can do so by passing a helper text component with that prop set to true with my current implementation that allows any ReactNode to be passed and rendered.

**Comment by wise-king-sullyman** [comment]
I would want to update the test suite to reflect the custom element passing functionality before merging, but I'm holding off on doing so until I'm able to get a 👍 on the general behavior/API here as it seemed like that was still a bit up in the air yesterday.

**Review by mcoker — APPROVED** [review]
LGTM! I do think the `string` type could be improved upon in case someone wants to include a span or like `Invalid date format. Date must be in the format of <i>mm/dd/yyyy</i>.` I imagine we'll also have other use cases where we want to render a component for the user in a basic use case, and allow them to compose it themselves for a more advanced use case (like we're doing here), and there may be more likely use cases there where the content passed isn't a string, and it may be good to have a consistent approach that works here and in the other places.

**Comment by wise-king-sullyman** [comment]
@mcoker yeah I would be more than happy to pass non-string content to the basic version of `HelperText` that we provide by default, but I couldn't think of a way to do that and still allow the fully custom option which didn't complicate the API at the same time (i.e. requiring an additional `isBasic` prop or something).

If anyone has ideas for that, or if you all think the approach I mention above would be worth the API complication tradeoff, please let me know!

**Review by thatblindgeye — COMMENTED** [review]
Looks good! I can holdoff on approving until you add tests and whatnot. Based on the original ask, it might be good to either add another Progress in the example showing how to use HelperText with a live region, or add an example description.

Regarding the comments above, I did do a little digging and came across the below StackOverflow post. Not entirely sure how viable/pertinent it would be for this, but if we do want to expand the logic as a followup at some point it could be worth looking into.

https://stackoverflow.com/a/69897991

File: packages/react-core/src/components/Progress/ProgressHelperText.tsx
Comment (tlabaj) [inline packages/react-core/src/components/Progress/ProgressHelperText.tsx:39]: Why do we not render this in a `HelperText` component? Is it so that the consumer can pass in a HelperText component themselves?

Comment (wise-king-sullyman) [reply packages/react-core/src/components/Progress/ProgressHelperText.tsx:39]: Correct, with this approach consumers can pass a `HelperText` (or multiple `HelperText`s) with custom options like `isLiveRegion` or a custom `variant`, or potentially different components rather than a `HelperText`.

Comment (tlabaj) [reply packages/react-core/src/components/Progress/ProgressHelperText.tsx:39]: @mcarrano What are you thought on having the "helperText be flexible? They way we have it implemented here, the consumer can pretty much have the helper the be anything.  If we are ok with the flexibility, I would suggest we not even have the component internally `HelperText` internally based on the string and just let the consumer pass it in.

Comment (mcarrano) [reply packages/react-core/src/components/Progress/ProgressHelperText.tsx:39]: I think I'm OK with the consumer passing this in as it allows for use cases we have not considered. We do have a Helper Text component: https://www.patternfly.org/v4/components/helper-text Does this leverage that? Is the concern that this does not cover all relevant use cases?

Comment (wise-king-sullyman) [reply packages/react-core/src/components/Progress/ProgressHelperText.tsx:39]: @mcarrano yes the current implementation allows consumers to pass a simple string and will automatically pass that string to a `HelperText` component and control the `HelperText` variant based on the variant of the `Progress`.

But it also allows the consumer to pass in a `HelperText` with a different configuration, or any other component they would like, if they have a use case for which the more automatic approach mentioned above doesn't work

Comment (mcarrano) [reply packages/react-core/src/components/Progress/ProgressHelperText.tsx:39]: That sounds fine to me. @tlabaj did you have concerns?

Comment (tlabaj) [reply packages/react-core/src/components/Progress/ProgressHelperText.tsx:39]: I actually think it is a little unexpected that only a string would cause the component to create the `HelperText` since the `HelperTextItem` type is a react node.

**Review by thatblindgeye — APPROVED** [review]
Looks good! The only nit I'd have is we could include a description for the example, though since the prop and example are both marked beta (and I can't really nail down *what* that description could be at the moment 😆 ) I wouldn't block over it.

File: packages/react-core/src/components/Progress/examples/ProgressHelperText.tsx
Comment (tlabaj) [inline packages/react-core/src/components/Progress/examples/ProgressHelperText.tsx:2]: Can we define a capitalize locally in the example.  The one you are importing her was really defined as an internal helper and the usage is not documented for the consumer. I have an issue open for breaking change release to update our exports to not include internal helpers.

**Review by tlabaj — CHANGES_REQUESTED** [review]
Looks good!  Just one comment.

**Review by mcoker — APPROVED** [review]
👍 nice!

**Review by tlabaj — APPROVED** [review]
LGTM

**Review by mmenestr — APPROVED** [review]

**Pull request merged**
